src/animation/AnimatedComponent.js

```diff
@@ -1,53 +1,58 @@
 /* @flow strict-local */
-import React, { PureComponent } from 'react';
+import React, { useRef, useCallback, useLayoutEffect } from 'react';
 import type { Node as React$Node } from 'react';
 import { Animated, Easing } from 'react-native';
 
+import { usePrevious } from '../reactUtils';
 import type { Style } from '../types';
 
 type Props = $ReadOnly<{|
   stylePropertyName: string,
   fullValue: number,
   children: React$Node,
   style?: Style,
-  visible: boolean,
-  useNativeDriver: boolean,
-  delay: number,
+  visible?: boolean,
+  useNativeDriver?: boolean,
+  delay?: number,
 |}>;
 
-export default class AnimatedComponent extends PureComponent<Props> {
-  static defaultProps = {
-    visible: true,
-    useNativeDriver: true,
-    delay: 0,
-  };
-
-  animatedValue = new Animated.Value(0);
-
-  animate() {
-    Animated.timing(this.animatedValue, {
-      toValue: this.props.visible ? this.props.fullValue : 0,
-      delay: this.props.delay,
+/**
+ * Animates the specified style property on visibility change.
+ */
+export default function AnimatedComponent(props: Props): React$Node {
+  const {
+    visible = true,
+    useNativeDriver = true,
+    delay = 0,
+    fullValue,
+    children,
+    stylePropertyName,
+    style,
+  } = props;
+
+  const prevVisible = usePrevious(visible);
+
+  const animatedValue = useRef(new Animated.Value(0));
+
+  const animate = useCallback(() => {
+    Animated.timing(animatedValue.current, {
+      toValue: visible ? fullValue : 0,
+      delay,
       duration: 300,
-      useNativeDriver: this.props.useNativeDriver,
+      useNativeDriver,
       easing: Easing.out(Easing.poly(4)),
     }).start();
-  }
+  }, [delay, fullValue, useNativeDriver, visible]);
```

Comment:
I think this is expected:

https://github.com/facebook/react/issues/16121#issuecomment-511369830
https://github.com/facebook/react/issues/14387#issuecomment-455415008
https://github.com/facebook/react/issues/14387#issuecomment-503616820

Reply:
Hmm, those comments explain that we shouldn't put `animatedValue.current` as a dependency, and I think I understand that.

But what I'd expect the linter to look at in the first place is the local variables that are getting closed over, so `animatedValue`, not any of its properties. In the first linked comment's terms, that's a ref container, not an example of `ref.current`.

It's actually just fine to not have it in the dependencies list, because it comes straight from a `useRef`, and that's guaranteed to give the same value on re-render. So having it as a dependency would never have any effect, because it'll never see a change.

I just wouldn't have expected the linter to detect that nuance, so I'd have expected it to flag this.

... Oh right, but this linter is from the React developers -- it's not something a random person decided to write. So maybe it's not so surprising that they did the work to capture that. And in fact it looks like [this](https://github.com/facebook/react/blob/321087d1343638e2d45e53971a7d8c384549940b/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L150-L161) (and the other mention of "useRef" a [bit lower down](https://github.com/facebook/react/blob/321087d1343638e2d45e53971a7d8c384549940b/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L223-L225)) is the pertinent bit of logic. The whole file is 1814 lines long -- so yeah, they weren't slacking.

Reply:
Interesting; yeah, it's nice to see that they took care writing that lint rule.